vm/src/stdlib/array.rs

```diff
@@ -440,7 +550,23 @@ impl PyArray {
         obj: PyObjectRef,
         vm: &VirtualMachine,
     ) -> PyResult<()> {
-        self.borrow_value_mut().setitem(needle, obj, vm)
+        match needle {
+            Either::A(i) => self.borrow_value_mut().setitem_by_index(i, obj, vm),
+            Either::B(slice) => {
+                // [obj] could ref to a PyObject same as self
+                // so before we call borrow_value_mut we have to clone the items from obj.payload
+                let items = match obj.payload::<PyArray>() {
+                    Some(array) => array.borrow_value().clone(),
```

Comment:
is this cloning really required? how about calling `setitem_by_slice` here?

```suggestion
                    Some(items) => self.borrow_value_mut().setitem_by_slice(slice, items, vm),
```